### ISSUE TITLE
SQL ERROR: "Prepared statement needs to be re-prepared

### ISSUE THREAD

**Issue by Nothing4You** [chris2511/xca#60]
Hi,

I've just tried to connect to a newly created DB in my MariaDB using xca on osx, however, it just fails with a popup like this: 
![screen shot 2018-08-27 at 14 51 51](https://user-images.githubusercontent.com/3694534/44660689-c072a800-aa08-11e8-8f6c-ae8068c68484.png)
Unfortunately I can't find any documentation explaining what I have to do.

According to https://hohnstaedt.de/xca/index.php it supports `MySQL (MariaDB)`, however, in the DB choice the MySQL option is `MySQL 3.x and 4.x`. Latest MariaDB is 10.3.7 while MySQL is at 8.0 (and 5.7 on the previous major version).

**Comment by chris2511** [comment]
Please try to put https://hohnstaedt.de/downloads/libmysqlclient.18.dylib into the /usr/lib/ folder on your Mac and see whether it works. For me it worked.
I need to investigate why the file hasn't been put into the xca.app bundle.
There is a new document describing the remote database use: https://hohnstaedt.de/xca/index.php/documentation/remote-databases
The  `MySQL 3.x and 4.x` term is copied from the Qt documentation and probably more irritating than helpful. I'll rename it to `MySQL and MariaDB`

**Comment by Nothing4You** [comment]
There is `xca.app/Contents/PlugIns/sqldrivers/libqsqlmysql.dylib` bundled.

Opening xca in a terminal gives the following output:
```
DB driver: "QSQLITE"
DB driver: "QMYSQL"
DB driver: "QMYSQL3"
DB driver: "QODBC"
DB driver: "QODBC3"
DB driver: "QPSQL"
DB driver: "QPSQL7"
Available Remote DB Drivers:  2
"QMYSQL3"
"QPSQL7"
Creating OID: "/Applications/xca.app/Contents/Resources/oids.txt" 8 1061 "1.3.6.1.4.1.311.20.2" "dom" "Domain Controller"
Creating OID: "/Applications/xca.app/Contents/Resources/oids.txt" 9 1062 "1.3.6.1.4.1.311.21.1" "MsCaV" "Microsoft CA Version"
Creating OID: "/Applications/xca.app/Contents/Resources/oids.txt" 11 1063 "1.3.6.1.4.1.311.10.3.4.1" "msEFSFR" "Microsoft EFS File Recovery"
Creating OID: "/Applications/xca.app/Contents/Resources/oids.txt" 12 1064 "1.3.6.1.5.5.8.2.2" "iKEIntermediate" "IP security end entity"
Creating OID: "/Applications/xca.app/Contents/Resources/oids.txt" 19 1065 "0.2.262.1.10.7.20" "nameDistinguisher" "Name distinguisher"
Creating OID: "/Applications/xca.app/Contents/Resources/oids.txt" 21 1066 "1.3.6.1.5.5.7.3.13" "id-kp-eapOverPPP" "EAP over PPP"
Creating OID: "/Applications/xca.app/Contents/Resources/oids.txt" 22 1067 "1.3.6.1.5.5.7.3.14" "id-kp-eapOverLAN" "EAP over Lan"
Opening database:
QFSFileEngine::open: No file name specified
Available Remote DB Drivers:  2
"QMYSQL3"
"QPSQL7"
OpenDb::getDbType:  "QMYSQL3"
OpenDb::getDbType:  "QMYSQL3"
QSqlDatabase: QMYSQL3 driver not loaded
QSqlDatabase: available drivers: QSQLITE QMYSQL QMYSQL3 QODBC QODBC3 QPSQL QPSQL7
SQL ERROR: "Driver not loaded Driver not loaded"
```

I'll see if I can put it in /usr/lib for testing that later.

**Comment by chris2511** [comment]
libqsqlmysql.dylib is the one requiring libmysqlclient.18.dylib and then saying "Driver not loaded" if not found. Printing it 2 times is a QT bug

**Comment by Nothing4You** [comment]
After connecting to the newly created DB it throws a few popups:
```
"QUERY: sql.cpp:76 (SELECT MAX(stamp) +1 from items) - Rows selected: 1"
"QUERY: sql.cpp:81 (UPDATE items SET stamp=? WHERE stamp=0[0]) - Rows affected: 0"
"QUERY: db_base.cpp:117 (SELECT * FROM view_public_keys) - Rows selected: -1"
SQL ERROR: "Prepared statement needs to be re-prepared QMYSQL3: Unable to execute statement - db_base.cpp:117 (SELECT * FROM view_public_keys)"
"QUERY: db_key.cpp:49 (SELECT pkey, COUNT(*) FROM x509super WHERE pkey IS NOT NULL GROUP by pkey) - Rows selected: 0"
"QUERY: db_base.cpp:117 (SELECT * FROM view_requests) - Rows selected: -1"
SQL ERROR: "Prepared statement needs to be re-prepared QMYSQL3: Unable to execute statement - db_base.cpp:117 (SELECT * FROM view_requests)"
"QUERY: db_base.cpp:117 (SELECT * FROM view_certs) - Rows selected: -1"
SQL ERROR: "Prepared statement needs to be re-prepared QMYSQL3: Unable to execute statement - db_base.cpp:117 (SELECT * FROM view_certs)"
"QUERY: db_base.cpp:117 (SELECT * FROM view_templates) - Rows selected: -1"
SQL ERROR: "Prepared statement needs to be re-prepared QMYSQL3: Unable to execute statement - db_base.cpp:117 (SELECT * FROM view_templates)"
"QUERY: db_base.cpp:117 (SELECT * FROM view_crls) - Rows selected: -1"
SQL ERROR: "Prepared statement needs to be re-prepared QMYSQL3: Unable to execute statement - db_base.cpp:117 (SELECT * FROM view_crls)"
```

**Comment by Nothing4You** [comment]
After adding a few items to the db using xca I can see the entries in the DB and the insert logs on my terminal, however, if I close and reopen the DB I get those same popups again and I can't see any entries from the last session.

Another question while at this topic, is parallel usage by multiple users supported? Judging from the lack of a refresh option I'm guessing it isn't?

**Comment by chris2511** [comment]
Can you tell me how to reproduce by starting from an empty database?
This looks like an XCA bug in conjunction with the MySQL driver.

Parallel usage is partly supported by transactions protected SQL statement groups  and
the "stamp" column in the items, incremented by each change and an indicator for additionally
connected XCA instances to recognize changes.
So currently you need to reconnect to see changes done by others.
Parallel changes however should be protected by the transactions already.

**Comment by Nothing4You** [comment]
Reproduction steps for me:
1. Create a new DB (in my case `xca2`)
2. Start xca
3. Connect to DB
4. see this (with a popup on every SQL error)
```
OpenDb::getDbType:  "QMYSQL3"
OpenDb::getDbType:  "QMYSQL3"
"MW_database.cpp(40) Transaction: Begin Level 1, E:0 "
table_prefix: ""
EXEC[0]: 'CREATE TABLE settings (key_ CHAR(20) UNIQUE, value VARCHAR(1024))'
EXEC[0]: 'INSERT INTO settings (key_, value) VALUES ('schema', '1')'
EXEC[0]: 'CREATE TABLE items(id INTEGER PRIMARY KEY, name VARCHAR(128), type INTEGER, source INTEGER, date CHAR(15), comment VARCHAR(2048))'
EXEC[0]: 'CREATE TABLE public_keys (item INTEGER, type CHAR(4), hash INTEGER, len INTEGER, public VARCHAR(10000), FOREIGN KEY (item) REFERENCES items (id))'
EXEC[0]: 'CREATE TABLE private_keys (item INTEGER, ownPass INTEGER, private VARCHAR(10000), FOREIGN KEY (item) REFERENCES items (id))'
EXEC[0]: 'CREATE TABLE tokens (item INTEGER, card_manufacturer VARCHAR(64), card_serial VARCHAR(64), card_model VARCHAR(64), card_label VARCHAR(64), slot_label VARCHAR(64), object_id VARCHAR(64), FOREIGN KEY (item) REFERENCES items (id))'
EXEC[0]: 'CREATE TABLE token_mechanism (item INTEGER, mechanism INTEGER, FOREIGN KEY (item) REFERENCES items (id))'
EXEC[0]: 'CREATE TABLE x509super (item INTEGER, subj_hash INTEGER, pkey INTEGER, key_hash INTEGER, FOREIGN KEY (item) REFERENCES items (id), FOREIGN KEY (pkey) REFERENCES items (id)) '
EXEC[0]: 'CREATE TABLE requests (item INTEGER, hash INTEGER, signed INTEGER, request VARCHAR(10000), FOREIGN KEY (item) REFERENCES items (id)) '
EXEC[0]: 'CREATE TABLE certs (item INTEGER, hash INTEGER, iss_hash INTEGER, serial VARCHAR(64), issuer INTEGER, ca INTEGER, cert VARCHAR(10000), FOREIGN KEY (item) REFERENCES items (id), FOREIGN KEY (issuer) REFERENCES items (id)) '
EXEC[0]: 'CREATE TABLE authority (item INTEGER, template INTEGER, crlExpire CHAR(15), crlNo INTEGER, crlDays INTEGER, dnPolicy VARCHAR(1024), FOREIGN KEY (item) REFERENCES items (id), FOREIGN KEY (template) REFERENCES items (id)) '
EXEC[0]: 'CREATE TABLE crls (item INTEGER, hash INTEGER, num INTEGER, iss_hash INTEGER, issuer INTEGER, crl VARCHAR(10000), FOREIGN KEY (item) REFERENCES items (id), FOREIGN KEY (issuer) REFERENCES items (id)) '
EXEC[0]: 'CREATE TABLE revocations (caId INTEGER, serial VARCHAR(64), date CHAR(15), invaldate CHAR(15), crlNo INTEGER, reasonBit INTEGER, FOREIGN KEY (caId) REFERENCES items (id))'
EXEC[0]: 'CREATE TABLE templates (item INTEGER, version INTEGER, template VARCHAR(10000), FOREIGN KEY (item) REFERENCES items (id))'
EXEC[1]: 'CREATE VIEW view_public_keys AS SELECT items.id, items.name, items.type AS item_type, items.date, items.source, items.comment, public_keys.type as key_type, public_keys.len, public_keys.public, private_keys.ownPass, tokens.card_manufacturer, tokens.card_serial, tokens.card_model, tokens.card_label, tokens.slot_label, tokens.object_id FROM public_keys LEFT JOIN items ON public_keys.item = items.id LEFT JOIN private_keys ON private_keys.item = public_keys.item LEFT JOIN tokens ON public_keys.item = tokens.item'
EXEC[1]: 'CREATE VIEW view_certs AS SELECT items.id, items.name, items.type, items.date AS item_date, items.source, items.comment, x509super.pkey, certs.serial AS certs_serial, certs.issuer, certs.ca, certs.cert, authority.template, authority.crlExpire, authority.crlNo AS auth_crlno, authority.crlDays, authority.dnPolicy, revocations.serial, revocations.date, revocations.invaldate, revocations.crlNo, revocations.reasonBit FROM certs LEFT JOIN items ON certs.item = items.id LEFT JOIN x509super ON x509super.item = certs.item LEFT JOIN authority ON authority.item = certs.item LEFT JOIN revocations ON revocations.caId = certs.issuer AND revocations.serial = certs.serial'
EXEC[1]: 'CREATE VIEW view_requests AS SELECT items.id, items.name, items.type, items.date, items.source, items.comment, x509super.pkey, requests.request, requests.signed FROM requests LEFT JOIN items ON requests.item = items.id LEFT JOIN x509super ON x509super.item = requests.item'
EXEC[1]: 'CREATE VIEW view_crls AS SELECT items.id, items.name, items.type, items.date, items.source, items.comment, crls.num, crls.issuer, crls.crl FROM crls LEFT JOIN items ON crls.item = items.id '
EXEC[1]: 'CREATE VIEW view_templates AS SELECT items.id, items.name, items.type, items.date, items.source, items.comment, templates.version, templates.template FROM templates LEFT JOIN items ON templates.item = items.id'
EXEC[1]: 'UPDATE settings SET value='2' WHERE key_='schema''
EXEC[2]: 'CREATE INDEX i_settings_key_ ON settings (key_)'
EXEC[2]: 'CREATE INDEX i_items_id ON items (id)'
EXEC[2]: 'CREATE INDEX i_public_keys_item ON public_keys (item)'
EXEC[2]: 'CREATE INDEX i_public_keys_hash ON public_keys (hash)'
EXEC[2]: 'CREATE INDEX i_private_keys_item ON private_keys (item)'
EXEC[2]: 'CREATE INDEX i_tokens_item ON tokens (item)'
EXEC[2]: 'CREATE INDEX i_token_mechanism_item ON token_mechanism (item)'
EXEC[2]: 'CREATE INDEX i_x509super_item ON x509super (item)'
EXEC[2]: 'CREATE INDEX i_x509super_subj_hash ON x509super (subj_hash)'
EXEC[2]: 'CREATE INDEX i_x509super_key_hash ON x509super (key_hash)'
EXEC[2]: 'CREATE INDEX i_x509super_pkey ON x509super (pkey)'
EXEC[2]: 'CREATE INDEX i_requests_item ON requests (item)'
EXEC[2]: 'CREATE INDEX i_requests_hash ON requests (hash)'
EXEC[2]: 'CREATE INDEX i_certs_item ON certs (item)'
EXEC[2]: 'CREATE INDEX i_certs_hash ON certs (hash)'
EXEC[2]: 'CREATE INDEX i_certs_iss_hash ON certs (iss_hash)'
EXEC[2]: 'CREATE INDEX i_certs_serial ON certs (serial)'
EXEC[2]: 'CREATE INDEX i_certs_issuer ON certs (issuer)'
EXEC[2]: 'CREATE INDEX i_certs_ca ON certs (ca)'
EXEC[2]: 'CREATE INDEX i_authority_item ON authority (item)'
EXEC[2]: 'CREATE INDEX i_crls_item ON crls (item)'
EXEC[2]: 'CREATE INDEX i_crls_hash ON crls (hash)'
EXEC[2]: 'CREATE INDEX i_crls_iss_hash ON crls (iss_hash)'
EXEC[2]: 'CREATE INDEX i_crls_issuer ON crls (issuer)'
EXEC[2]: 'CREATE INDEX i_revocations_caId_serial ON revocations (caId, serial)'
EXEC[2]: 'CREATE INDEX i_templates_item ON templates (item)'
EXEC[2]: 'UPDATE settings SET value='3' WHERE key_='schema''
EXEC[3]: 'CREATE VIEW view_private AS SELECT name, private FROM private_keys JOIN items ON items.id = private_keys.item'
EXEC[3]: 'UPDATE settings SET value='4' WHERE key_='schema''
EXEC[4]: 'ALTER TABLE items ADD stamp INTEGER NOT NULL DEFAULT 0'
EXEC[4]: 'CREATE INDEX i_items_stamp ON items (stamp)'
EXEC[4]: 'UPDATE settings SET value='5' WHERE key_='schema''
EXEC[5]: 'ALTER TABLE settings RENAME TO __settings'
EXEC[5]: 'CREATE TABLE settings (key_ CHAR(20) UNIQUE, value VARCHAR(10000))'
EXEC[5]: 'INSERT INTO settings(key_, value) SELECT key_, value FROM __settings'
EXEC[5]: 'DROP TABLE __settings'
EXEC[5]: 'UPDATE settings SET value='6' WHERE key_='schema''
"MW_database.cpp(52) Transaction: Commit Level 0, E:0 "
"QUERY: sql.cpp:76 (SELECT MAX(stamp) +1 from items) - Rows selected: 1"
"QUERY: sql.cpp:81 (UPDATE items SET stamp=? WHERE stamp=0[0]) - Rows affected: 0"
OpenDb::getDbType:  "QMYSQL3"
OpenDb::getDbType:  "QMYSQL3"
DB-DESC: "user@dbhost/QMYSQL3:xca2" "user@dbhost/QMYSQL3:xca2" QSqlError("", "", "")
"settings.cpp(110) Transaction: Begin Level 1, E:0 "
"QUERY: settings.cpp:116 (INSERT INTO settings (value, key_) VALUES (?,?)[hash, pwhash]) - Rows affected: 1"
"settings.cpp(123) Transaction: Commit Level 0, E:0 "
"QUERY: sql.cpp:76 (SELECT MAX(stamp) +1 from items) - Rows selected: 1"
"QUERY: sql.cpp:81 (UPDATE items SET stamp=? WHERE stamp=0[0]) - Rows affected: 0"
"QUERY: db_base.cpp:117 (SELECT * FROM view_public_keys) - Rows selected: -1"
SQL ERROR: "Prepared statement needs to be re-prepared QMYSQL3: Unable to execute statement - db_base.cpp:117 (SELECT * FROM view_public_keys)"
"QUERY: db_key.cpp:49 (SELECT pkey, COUNT(*) FROM x509super WHERE pkey IS NOT NULL GROUP by pkey) - Rows selected: 0"
"QUERY: db_base.cpp:117 (SELECT * FROM view_requests) - Rows selected: -1"
SQL ERROR: "Prepared statement needs to be re-prepared QMYSQL3: Unable to execute statement - db_base.cpp:117 (SELECT * FROM view_requests)"
"QUERY: db_base.cpp:117 (SELECT * FROM view_certs) - Rows selected: -1"
SQL ERROR: "Prepared statement needs to be re-prepared QMYSQL3: Unable to execute statement - db_base.cpp:117 (SELECT * FROM view_certs)"
"QUERY: db_base.cpp:117 (SELECT * FROM view_templates) - Rows selected: -1"
SQL ERROR: "Prepared statement needs to be re-prepared QMYSQL3: Unable to execute statement - db_base.cpp:117 (SELECT * FROM view_templates)"
"QUERY: db_base.cpp:117 (SELECT * FROM view_crls) - Rows selected: -1"
SQL ERROR: "Prepared statement needs to be re-prepared QMYSQL3: Unable to execute statement - db_base.cpp:117 (SELECT * FROM view_crls)"
```

**Comment by vsc55** [comment]
hello, in windows the same error also occurs.

![image](https://user-images.githubusercontent.com/13438676/54920393-6b960200-4f03-11e9-9163-1312b43c60a6.png)

Version Server: 10.1.37-MariaDB-0+deb9u1 - Raspbian 9.0

**Comment by yakatz** [comment]
@vsc55 I suddenly started getting the same error on Windows after it had been working for a while. I found that restarting the MariaDB server fixed the issue.

**Comment by yakatz** [comment]
It's great when you set up a new computer and have an issue, then you find that you solved the issue yourself and have no recollection of doing it :-)

**Comment by reetp** [comment]
> @vsc55 I suddenly started getting the same error on Windows after it had been working for a while. I found that restarting the MariaDB server fixed the issue.

I have been testing this myself with various packages on xUbuntu 22.04

Exact same issue. After creating the initial CA I logged back in and couldn't see it. Same errors.

In desperation I restarted the remote Maria 10.5.x server and it worked :worried: 

Makes no sense at all and have never seen it happen before (hence why I never tried a restart)

I also can'r replicate it!
